go-controller/pkg/clustermanager/node_network_controller.go

```diff
@@ -0,0 +1,178 @@
+package clustermanager
+
+import (
+	"fmt"
+	"net"
+	"sync"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/clustermanager/subnetallocator"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/kube"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+	kapi "k8s.io/api/core/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+)
+
+// NodeNetworkController structure manages the network allocations for a 'network'
+// for each node. Presently it only manages the subnet allocation.
+type NodeNetworkController struct {
+	kube         kube.Interface
+	watchFactory *factory.WatchFactory
+
+	networkName            string
+	clusterSubnetAllocator *subnetallocator.HostSubnetAllocator
+
+	subnetAllocatorLock sync.Mutex
+}
+
+func newNodeNetworkController(kube kube.Interface, wf *factory.WatchFactory,
+	networkName string) *NodeNetworkController {
+	return &NodeNetworkController{
+		kube:                   kube,
+		watchFactory:           wf,
+		networkName:            networkName,
+		clusterSubnetAllocator: subnetallocator.NewHostSubnetAllocator(),
+	}
+}
+
+func (nnc *NodeNetworkController) InitSubnetAllocatorRanges(subnets []config.CIDRNetworkEntry) error {
+	return nnc.clusterSubnetAllocator.InitRanges(subnets)
+}
+
+func (nnc *NodeNetworkController) syncNodes(nodes []interface{}) error {
+	nnc.subnetAllocatorLock.Lock()
+	defer func() {
+		nnc.subnetAllocatorLock.Unlock()
+	}()
+
+	for _, tmp := range nodes {
+		node, ok := tmp.(*kapi.Node)
+		if !ok {
+			return fmt.Errorf("spurious object in syncNodes: %v", tmp)
+		}
+
+		if util.NoHostSubnet(node) {
+			continue
+		}
+
+		hostSubnets, _ := util.ParseNodeHostSubnetAnnotation(node, nnc.networkName)
+		klog.V(5).Infof("Node %s contains subnets: %v for network : %s", node.Name, hostSubnets, nnc.networkName)
+		if err := nnc.clusterSubnetAllocator.MarkSubnetsAllocated(node.Name, hostSubnets...); err != nil {
+			utilruntime.HandleError(err)
+		}
+	}
+
+	return nil
+}
+
+func (nnc *NodeNetworkController) addUpdateNode(node *kapi.Node) error {
+	nnc.subnetAllocatorLock.Lock()
+	defer func() {
+		nnc.subnetAllocatorLock.Unlock()
+	}()
+
+	existingSubnets, err := util.ParseNodeHostSubnetAnnotation(node, nnc.networkName)
+	if err != nil && !util.IsAnnotationNotSetError(err) {
+		// Log the error and try to allocate new subnets
+		klog.Infof("Failed to get node %s host subnets annotations for network %s : %v", node.Name, nnc.networkName, err)
+
+	}
+
+	hostSubnets, allocatedSubnets, err := nnc.clusterSubnetAllocator.AllocateNodeSubnets(node.Name, existingSubnets, config.IPv4Mode, config.IPv6Mode)
+	if err != nil {
+		return err
+	}
+
+	if len(allocatedSubnets) == 0 {
+		return nil
+	}
+
+	// Release the allocation on error
+	defer func() {
+		if err != nil {
```

Comment:
This will never be true right? Below we do not set result of `updateNodeAnnotationWithRetry` to var `err`.